strawberry/channels/testing.py

```diff
@@ -0,0 +1,102 @@
+import dataclasses
+import uuid
+from typing import AsyncIterator, Dict, List, Optional, Tuple, Union
+
+from graphql import GraphQLError
+
+from channels.testing.websocket import WebsocketCommunicator
+from strawberry.subscriptions import GRAPHQL_TRANSPORT_WS_PROTOCOL
+from strawberry.subscriptions.protocols.graphql_transport_ws.types import (
+    ConnectionAckMessage,
+    ConnectionInitMessage,
+    SubscribeMessage,
+    SubscribeMessagePayload,
+)
+from strawberry.types import ExecutionResult
+
+
+class GraphQLWebsocketCommunicator(WebsocketCommunicator):
+    """
+    Usage:
+    ```python
+    import pytest
+    from strawberry.channels.testing import GqlWsCommunicator
+    from myapp.asgi import application
+
+
+    @pytest.fixture
+    async def gql_communicator():
+        async with GqlWsCommunicator(application, path="/graphql") as client:
+            yield client
+
+
+    async def test_subscribe_echo(gql_communicator):
+        async for res in gql_communicator.subscribe(
+            query='subscription { echo(message: "Hi") }'
+        ):
+            assert res.data == {"echo": "Hi"}
+    ```
+    """
+
+    def __init__(
+        self,
+        application,
+        path: str,
+        headers: Optional[List[Tuple[bytes, bytes]]] = None,
+        subprotocols: Optional[List[str]] = None,
+    ):
+        subprotocols = (
+            subprotocols
+            if subprotocols
+            else [
+                GRAPHQL_TRANSPORT_WS_PROTOCOL,
+            ]
+        )
+        super().__init__(application, path, headers, subprotocols)
+
+    async def __aenter__(self) -> "GraphQLWebsocketCommunicator":
+        await self.gql_init()
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
+        await self.disconnect()
+
+    async def gql_init(self):
+        res = await self.connect()
+        assert res == (True, GRAPHQL_TRANSPORT_WS_PROTOCOL)
+        await self.send_json_to(ConnectionInitMessage().as_dict())
+        response = await self.receive_json_from()
+        assert response == ConnectionAckMessage().as_dict()
+
+    async def subscribe(
+        self, query: str, variables: Optional[Dict] = None
+    ) -> Union[ExecutionResult, AsyncIterator[ExecutionResult]]:
+        id_ = uuid.uuid4().hex
+        await self.send_json_to(
+            SubscribeMessage(
+                id=id_,
+                payload=SubscribeMessagePayload(query=query, variables=variables),
+            ).as_dict()
+        )
+        while True:
+            response = await self.receive_json_from(timeout=5)
+            message_type = response["type"]
+            if message_type == "next":
```

Comment:
I would prefer if we used the message type constants defined in the subscriptions module instead of inline strings. It's for consitency and maintainability :)

Reply:
TBH I couldn't find any constants, though I can add if you want.